server/application-server/src/main/java/de/tum/in/www1/hephaestus/leaderboard/LeaguePointsCalculationService.java

```diff
@@ -0,0 +1,89 @@
+package de.tum.in.www1.hephaestus.leaderboard;
+
+import de.tum.in.www1.hephaestus.gitprovider.pullrequest.PullRequest;
+import de.tum.in.www1.hephaestus.gitprovider.user.User;
+import java.time.OffsetDateTime;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+@Service
+public class LeaguePointsCalculationService {
+    private final Logger logger = LoggerFactory.getLogger(LeaguePointsCalculationService.class);
+
+    public int calculateNewPoints(User user, LeaderboardEntryDTO entry) {
+        // Initialize new players with 1000 points
+        if (user.getLeaguePoints() == 0) {
+            user.setLeaguePoints(1000);
+        }
+
+        int oldPoints = user.getLeaguePoints();
+
+        double kFactor = getKFactor(user);
+        // Base decay
+        int decay = calculateDecay(user.getLeaguePoints());
+        // Bonus based on leaderboard score
+        int performanceBonus = calculatePerformanceBonus(entry.score());
+        // Additional bonus for placements
+        int placementBonus = calculatePlacementBonus(entry.rank());
+        // Calculate final point change
+        int pointChange = (int) (kFactor * (performanceBonus + placementBonus - decay));
+        // Apply minimum change to prevent extreme swings
+        int newPoints = Math.max(0, oldPoints + pointChange);
+        
+        logger.info("Points calculation: old={}, k={} decay={}, performanceBonus={}, placement={}, pointchange={}, new={}", 
+            oldPoints, kFactor, decay, performanceBonus, placementBonus, pointChange, newPoints);
+        
+        return newPoints;
+    }
+
+    /**
+     * Calculate the K factor for the user based on their current points and if they are a new player.
+     * @param user
+     * @return
+     * @see <a href="https://en.wikipedia.org/wiki/Elo_rating_system#Most_accurate_K-factor">Wikipedia: Most accurate K-factor</a>
+     */
+    private double getKFactor(User user) {
+        if (isNewPlayer(user)) {
+            return 2.0;
+        } else if (user.getLeaguePoints() < 1400) {
+            return 1.5;
+        } else if (user.getLeaguePoints() < 1800) {
+            return 1.2;
+        } else {
+            return 1.1;
+        }
+    }
+
+    /**
+     * Check if the user's earliest merged pull request is within the last 30 days.
+     * @param user
+     * @return
+     */
+    private boolean isNewPlayer(User user) {
+        return user.getMergedPullRequests().stream()
+            .filter(PullRequest::isMerged)
+            .map(PullRequest::getMergedAt)
+            .anyMatch(date -> date.isAfter(OffsetDateTime.now().minusDays(30)));
```

Comment:
is this condition not true if there is an old PR so a old Player?

Reply:
Could you elaborate on what you mean by that?
What the function is supposed to do is determine if the user has opened a PR (which has been merged) at least 30 days ago, meaning that the user has been "contributing" for at least a month as a criterium to consider them not a new player anymore.

Reply:
from my understanding, we get a match if there is a PR older than 30 days, right? But this would be isOldPlayer() check I think. so we need to adjust the condition or naming

Reply:
fixed in e3e1212.